src/components/ResizeDetector.js

```diff
@@ -0,0 +1,204 @@
+    if (shouldCancel) {
+      // cancel debounced handler
+      this.resizeHandler.cancel();
+      this.resizeHandler = null;
+    }
+  }
+
+  rafClean = () => {
+    const shouldClean = !!(this.raf && this.raf.cancel);
+    if (shouldClean) {
+      this.raf.cancel();
+      this.raf = null;
+    }
+  }
+
+  toggleObserver = (isOn) => {
+    const element = this.getElement();
+    if (!element) return false;
+
+    const type = isOn ? 'observe' : 'unobserve';
+    const handle = this.resizeObserver[type];
+    if (!handle) return false;
+
+    this.resizeObserver[type](element);
+    return true;
+  }
+
+  getElement = () => {
+    const { querySelector } = this.props;
+
+    const selectedElement = querySelector && document.querySelector(querySelector);
+
+    // eslint-disable-next-line react/no-find-dom-node
+    const currentElement = findDOMNode(this.element);
+    const parentElement = currentElement && currentElement.parentElement;
+
+    const element = selectedElement || parentElement;
+    return element;
+  };
+
+  createUpdater = () => {
+    this.rafClean();
+
+    this.raf = rafSchd(({ width, height }) => {
+      const { onResize } = this.props;
+
+      if (isFunction(onResize)) {
+        onResize(width, height);
+      }
+      if (this.unmounted) {
+        this.setState({ width, height });
+      }
+    });
+
+    return this.raf;
+  }
+
+  createResizeHandler = (entries) => {
+    const {
+      width: widthCurrent,
+      height: heightCurrent,
+    } = this.state;
+    const {
+      handleWidth,
+      handleHeight,
+    } = this.props;
+
+    const updater = this.createUpdater();
+
+    entries.forEach((entry) => {
+      const { width, height } = (entry && entry.contentRect) || {};
+
+      const isWidthChanged = handleWidth && widthCurrent !== width;
+      const isHeightChanged = handleHeight && heightCurrent !== height;
+      const isSizeChanged = isWidthChanged || isHeightChanged;
+
+      const isHaveWindow = checkIsHaveWindow();
+
+      const shouldSetSize = !this.skipOnMount && isSizeChanged && isHaveWindow;
+      if (shouldSetSize) {
+        updater({ width, height });
+      }
+
+      this.skipOnMount = false;
+    });
+  };
+
+  onRef = (el) => {
+    this.element = el;
+  };
+
+  render() {
+    const { width, height } = this.state;
+    const { render, children } = this.props;
+
+    const childProps = { width, height };
+    const isRenderProps = isFunction(render);
+    const isFunctional = isFunction(children);
+    const isComponent = isValidElement(children);
+
+    let component;
+    if (!component && isRenderProps) {
+      component = cloneElement(render(childProps), { key: 'resize-detector' });
+    }
+
+    if (!component && isFunctional) {
+      component = cloneElement(children(childProps));
+    }
+
+    if (!component && isComponent) {
+      component = cloneElement(children, childProps);
+    }
+
+    return <Reference ref={this.onRef}>{component || <div />}</Reference>;
```

Comment:
let's remove `<div />` from here.

It's ok when a component is `undefined` (when we use Callback pattern, right?)

If we pass `undefined` as a child to this component - then `this.element` will be undefined, right?

Just add another condition here
```
const currentElement = this.element && findDOMNode(this.element);
```

and everything will work as expected. But we will get rid of this extra `div` element

Reply:
Because I don't want to make the package more complicated! Adding a tag when using a callback will not break the interface!